Add observability quick start (#632)
langchain-ai/langsmith-docs: docs/observability/index.mdx

---
sidebar_label: Quick Start
sidebar_position: 0
table_of_contents: true
---

import {
  CodeTabs,
  python,
  typescript,
  ShellBlock,
} from "@site/src/components/InstructionsWithCode";
import { RegionalUrl } from "@site/src/components/RegionalUrls";

# Observability Quick Start

> This tutorial will get you up and running with our observability SDK by showing you how to

Reviewer: We should have 3 top level sections here: 

Get Started with LangSmith if you're using [LangChain](https://python.langchain.com/docs/introduction/)

LangChain integrates seamlessly with LangSmith, with no extra instrumentation needed. Learn how to start tracing with [LangChain](/observability/how_to_guides/tracing/trace_with_langchain). 

Get Started with LangSmith if you're using [LangGraph](https://langchain-ai.github.io/langgraph/)

LangGraph integrates seamlessly with LangSmith, with no extra instrumentation needed. Learn how to start tracing with [LangGraph](/observability/how_to_guides/tracing/trace_with_langgraph). 

Get Started instrumenting your application with with LangSmith 

<rest of your guide here>

Author: Having 3 headers with only a sentence under each one does not render well. I added a note instead to achieve the same purpose.

Reviewer: sg!

trace your application to LangSmith.

If you're already familiar with the observability SDK, or are interested in tracing more than just

> LLM calls you can skip to the [Next steps section](#next-steps),

Reviewer: ```suggestion
LLM calls you can skip to the [next steps section](#next-steps),
```

or check out the [how-to guides](../observability/how_to_guides).

## 1. Install Dependencies

<CodeTabs
  tabs={[
    {
      value: "python",
      label: "Python",
      language: "bash",
      content: `pip install -U langsmith openai`,
    },
    {
      value: "typescript",
      label: "TypeScript",
      language: "bash",
      content: `yarn add langsmith openai`,
    },
  ]}
  groupId="client-language"
/>

## 2. Create an API key

> To create an API key head to the <RegionalUrl text='Settings page' suffix='/settings' />. Then click **Create API Key.**

Reviewer: ```suggestion
To create an API key head to the <RegionalUrl text='LangSmith settings page' suffix='/settings' />. Then click **Create API Key.**
```

## 3. Set up your environment

<CodeTabs
  tabs={[

>     ShellBlock(`export LANGCHAIN_TRACING_V2=true

Reviewer: these env vars need to be changed to LANGSMITH_*

export LANGCHAIN_API_KEY="<your-langchain-api-key>"
# The example uses OpenAI, but it's not necessary if your code uses another LLM provider
export OPENAI_API_KEY="<your-openai-api-key>"`),
  ]}
  groupId="client-language"
/>

## 4. Define your application

> We will use a simple [RAG](https://www.mckinsey.com/featured-insights/mckinsey-explainers/what-is-retrieval-augmented-generation-rag)

Reviewer: ```suggestion
We will instrument a simple [RAG](https://www.mckinsey.com/featured-insights/mckinsey-explainers/what-is-retrieval-augmented-generation-rag)
```

application for this tutorial, but feel free to use your own code if you'd like - just make sure
it has an LLM call!

<details>

>   <summary>Application Code (Python and TypeScript)</summary>

Reviewer: ```suggestion
  <summary>Application Code</summary>
```

                  {"role": "user", "content": question},
              ],
              model="gpt-4o-mini",
          )
    `,
      typescript({ label: "TypeScript" })`
      import { OpenAI } from "openai";
      
      const openAIClient = new OpenAI();
      
      // This is the retriever we will use in RAG
      // This is mocked out, but it could be anything we want
      async function retriever(query: string) {
        return ["This is a document"];
      }
      
      // This is the end-to-end RAG chain.
      // It does a retrieval step then calls OpenAI
      async function rag(question: string) {
        const docs = await retriever(question);
        
        const systemMessage =
          "Answer the users question using only the provided information below:\\n\\n" +
          docs.join("\\n");
          
        return await openAIClient.chat.completions.create({
          messages: [
            { role: "system", content: systemMessage },
            { role: "user", content: question },
          ],
          model: "gpt-4o-mini",
        });
      }
    `,
    ]}
  />
</details>

## 5. Trace OpenAI calls

> The first thing you might want to trace is all your OpenAI calls. LangSmith makes this easy with the `wrap_openai/wrapOpenAI` wrappers.

Reviewer: add links to SDK docs

https://docs.smith.langchain.com/reference/python/wrappers/langsmith.wrappers._openai.wrap_openai
https://docs.smith.langchain.com/reference/js/functions/wrappers_openai.wrapOpenAI

    `,
    typescript({ label: "TypeScript" })`
      import { OpenAI } from "openai";
      // highlight-next-line
      import { wrapOpenAI } from "langsmith/wrappers";
      
      // highlight-next-line
      const openAIClient = wrapOpenAI(new OpenAI());
      
      // This is the retriever we will use in RAG
      // This is mocked out, but it could be anything we want
      async function retriever(query: string) {
        return ["This is a document"];
      }
      
      // This is the end-to-end RAG chain.
      // It does a retrieval step then calls OpenAI
      async function rag(question: string) {
        const docs = await retriever(question);
        
        const systemMessage =
          "Answer the users question using only the provided information below:\\n\\n" +
          docs.join("\\n");
          
        return await openAIClient.chat.completions.create({
          messages: [
            { role: "system", content: systemMessage },
            { role: "user", content: question },
          ],
          model: "gpt-4o-mini",
        });
      }
    `,
  ]}
/>

Now when you call your application as follows:

```python
rag("where did harrison work")
```

> This will produce a trace of just the OpenAI call - it should look something like [this](https://smith.langchain.com/public/e7b7d256-10fe-4d49-a8d5-36ca8e5af0d2/r)

Reviewer: ```suggestion
This will produce a trace of just the OpenAI call in LangSmith's default tracing project. It should look something like [this](https://smith.langchain.com/public/e7b7d256-10fe-4d49-a8d5-36ca8e5af0d2/r).
```

![](./tutorials/static/tracing_tutorial_openai.png)

## 6. Trace entire application

> You can also use the `traceable` decorator to trace your entire application instead of just the LLM calls.

Reviewer: add docs links

https://langsmith-docs-bdk0fivr6-langchain.vercel.app/reference/js/functions/traceable.traceable (why does the URL change)
https://docs.smith.langchain.com/reference/python/run_helpers/langsmith.run_helpers.traceable

      const rag = traceable(async function rag(question: string) {
        const docs = await retriever(question);
        
        const systemMessage =
          "Answer the users question using only the provided information below:\\n\\n" +
          docs.join("\\n");
          
        return await openAIClient.chat.completions.create({
          messages: [
            { role: "system", content: systemMessage },
            { role: "user", content: question },
          ],
          model: "gpt-4o-mini",
        });
      });
    `,
  ]}
/>

Now if you call your application as follows:

```python
rag("where did harrison work")
```

This will produce a trace of just the entire pipeline (with the OpenAI call as a child run) - it should look something like [this](https://smith.langchain.com/public/2174f4e9-48ab-4f9e-a8c4-470372d976f1/r)

![](./tutorials/static/tracing_tutorial_chain.png)

## Next steps

Congratulations! If you've made it this far, you're well on your way to being an expert in observability with LangSmith.
Here are some topics you might want to explore next:

- [Trace a `LangChain` application](./observability/how_to_guides/tracing/trace_with_langchain)

> - [Trace a `LangGraph` application](./observability/how_to_guides/tracing/trace_with_langgraph)

Reviewer: remove these since we moved them up

- [Trace multiturn conversations](./observability/how_to_guides/monitoring/threads)

> - [Send traces to a specific project](./observability/how_to_guides/tracing/log_traces_to_project)

Reviewer: /observability/how_to_guides/monitoring/filter_traces_in_application

this could be a good one for next steps too

                  {"role": "user", "content": question},
              ],
              model="gpt-4o-mini",
          )
    `,
      typescript({ label: "TypeScript" })`
      import { OpenAI } from "openai";
      
      const openAIClient = new OpenAI();
      
      // This is the retriever we will use in RAG
      // This is mocked out, but it could be anything we want
      async function retriever(query: string) {
        return ["This is a document"];
      }
      
      // This is the end-to-end RAG chain.
      // It does a retrieval step then calls OpenAI
      async function rag(question: string) {
        const docs = await retriever(question);
        
        const systemMessage =
          "Answer the users question using only the provided information below:\\n\\n" +
          docs.join("\\n");
          
        return await openAIClient.chat.completions.create({
          messages: [
            { role: "system", content: systemMessage },
            { role: "user", content: question },
          ],
          model: "gpt-4o-mini",
        });
      }
    `,
    ]}
  />
</details>

## 5. Trace OpenAI calls

> The first thing you might want to trace is all your OpenAI calls. LangSmith makes this easy with the [`wrap_openai`](https://docs.smith.langchain.com/reference/python/wrappers/langsmith.wrappers._openai.wrap_openai_)/[`wrapOpenAI`](https://docs.smith.langchain.com/reference/js/functions/wrappers_openai.wrapOpenAI) wrappers.

Reviewer: ```suggestion
The first thing you might want to trace is all your OpenAI calls. LangSmith makes this easy with the [`wrap_openai`](https://docs.smith.langchain.com/reference/python/wrappers/langsmith.wrappers._openai.wrap_openai_) (Python) or [`wrapOpenAI`](https://docs.smith.langchain.com/reference/js/functions/wrappers_openai.wrapOpenAI) (TypeScript) wrappers.
```

      // This is mocked out, but it could be anything we want
      async function retriever(query: string) {
        return ["This is a document"];
      }
      
      // This is the end-to-end RAG chain.
      // It does a retrieval step then calls OpenAI
      async function rag(question: string) {
        const docs = await retriever(question);
        
        const systemMessage =
          "Answer the users question using only the provided information below:\\n\\n" +
          docs.join("\\n");
          
        return await openAIClient.chat.completions.create({
          messages: [
            { role: "system", content: systemMessage },
            { role: "user", content: question },
          ],
          model: "gpt-4o-mini",
        });
      }
    `,
  ]}
/>

Now when you call your application as follows:

```python
rag("where did harrison work")
```

This will produce a trace of just the OpenAI call in LangSmith's default tracing project. It should look something like [this](https://smith.langchain.com/public/e7b7d256-10fe-4d49-a8d5-36ca8e5af0d2/r).

![](./tutorials/static/tracing_tutorial_openai.png)

## 6. Trace entire application

> You can also use the [`traceable`](https://docs.smith.langchain.com/reference/python/run_helpers/langsmith.run_helpers.traceable)/[`traceable`](https://langsmith-docs-bdk0fivr6-langchain.vercel.app/reference/js/functions/traceable.traceable) decorator to trace your entire application instead of just the LLM calls.

Reviewer: ```suggestion
You can also use the [`traceable`] decorator ([Python](https://docs.smith.langchain.com/reference/python/run_helpers/langsmith.run_helpers.traceable) or [TypeScript](https://langsmith-docs-bdk0fivr6-langchain.vercel.app/reference/js/functions/traceable.traceable)) to trace your entire application instead of just the LLM calls.
```
